### PR TITLE
feat(core): Can access component optional props

### PR DESCRIPTION
**Pull request by BatuhanW**
Refactored CanAccess component to work without any props.

Screenshot 2023-03-16 at 16 54 34

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

_Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request._

Please provide enough information so that others can review your pull request:

Explain the **details** for making this change. What existing problem does the pull request solve?

### Test plan (required)

Demonstrate the code is solid. If not, please add `WIP:` in its title.

### Closing issues

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### Self Check before Merge

Please check all items below before review.

-   [ ] Corresponding issues are created/updated or not needed
-   [ ] Docs are updated/provided or not needed
-   [ ] Examples are updated/provided or not needed
-   [ ] TypeScript definitions are updated/provided or not needed
-   [ ] Tests are updated/provided or not needed
-   [ ] Changesets are provided or not needed

### CODE REVIEW

File: packages/core/src/components/canAccess/index.spec.tsx [hunk 1]
```diff
@@ -4,92 +4,444 @@ import { render, TestWrapper } from "@test";
...
-                            if (action === "access" && resource === "posts") {
-                                return Promise.resolve({ can: true });
-                            }
-                            return Promise.resolve({ can: false });
+    describe("when props are passed", () => {
+        describe("when new router", () => {
+            describe("when pick resource undefined", () => {
+                describe("when fallback is empty", () => {
+                    const { container } = render(
+                        <CanAccess action="access" resource="posts">
```
Comment (omeraplak) [inline packages/core/src/components/canAccess/index.spec.tsx:19]:
```suggestion
                        <CanAccess action="list" resource="posts">
```

File: packages/core/src/components/canAccess/index.spec.tsx [hunk 2]
```diff
@@ -4,92 +4,444 @@ import { render, TestWrapper } from "@test";
...
+                            Accessible
+                        </CanAccess>,
+                        {
+                            wrapper: TestWrapper({
+                                accessControlProvider: {
+                                    can: ({ resource, action }) => {
+                                        if (
+                                            action === "access" &&
```
Comment (omeraplak) [inline packages/core/src/components/canAccess/index.spec.tsx:27]:
```suggestion
                                            action === "list" &&
```

**Review by alicanerdurmaz — APPROVED** [review]

File: documentation/docs/packages/documentation/routers/remix-router.md [hunk 3]
```diff
@@ -868,6 +775,202 @@ export const loader: LoaderFunction = async ({ params, request, context }) => {
...
+Tadaa! that's all! 🎉
+
+### Client Side
+
+For client-side, you can wrap your pages with [`CanAccess`](/docs/api-reference/core/components/accessControl/can-access) component from `@refinedev/core` to protect your pages from unauthorized access.
+
+## FAQ
+
+#
+
```
Comment (aliemir) [inline documentation/docs/packages/documentation/routers/remix-router.md:885]:
```suggestion
```

File: documentation/docs/packages/documentation/routers/remix-router.md [hunk 4]
```diff
@@ -868,6 +775,202 @@ export const loader: LoaderFunction = async ({ params, request, context }) => {
...
+    return json({ initialData: data });
+}
+```
+
+Tadaa! that's all! 🎉
+
+### Client Side
+
+For client-side, you can wrap your pages with [`CanAccess`](/docs/api-reference/core/components/accessControl/can-access) component from `@refinedev/core` to protect your pages from unauthorized access.
```
Comment (aliemir) [inline documentation/docs/packages/documentation/routers/remix-router.md:880]: We can say "wrap your pages/components" and share a tiny snippet that wraps the `<CreateButton>` with `<CanAccess>` maybe. @omeraplak @BatuhanW

File: documentation/docs/packages/documentation/routers/remix-router.md [hunk 5]
```diff
@@ -868,6 +775,202 @@ export const loader: LoaderFunction = async ({ params, request, context }) => {
...
+
+    return (
+        <List>
+            <Table {...tableProps} rowKey="id">
+                <Table.Column dataIndex="id" title="ID" />
+                <Table.Column dataIndex="status" title="Status" />
+                <Table.Column dataIndex="title" title="Title" />
+            </Table>
+        </List>
+    );
```
Comment (aliemir) [inline documentation/docs/packages/documentation/routers/remix-router.md:850]: I don't think we need this much details in the component, takes away the attention from the `loader`

File: documentation/docs/packages/documentation/routers/nextjs-router.md [hunk 6]
```diff
@@ -793,6 +700,175 @@ export const getServerSideProps = async (context) => {
...
+        return { can: true };
+    },
+};
+```
+
+Then, let's create our posts page.
+
+```tsx title="pages/posts/index.tsx"
+import { accessControlProvider } from "src/accessControlProvider";
+
```
Comment (aliemir) [inline documentation/docs/packages/documentation/routers/nextjs-router.md:752]: On the contrary to the remix's access control section, here we have no info about the page 😅 Maybe we can just do

```tsx
export default function PostList() { /* ... */ }
```

File: documentation/docs/packages/documentation/routers/nextjs-router.md [hunk 7]
```diff
@@ -793,6 +700,175 @@ export const getServerSideProps = async (context) => {
 
 This needs to be done for all the routes that we want to protect.
 
+## Access Control
+
+In Next.js you can achieve access control in multiple ways;
+
+On the client-side you can wrap your pages with `CanAccess` component from `@refinedev/core` to protect your pages from unauthorized access.
+
+And on the server-side you can use your `accessControlProvider`'s `can` function inside server side functions (getServerSideProps) to redirect unauthorized users to other pages..
```
Comment (aliemir) [inline documentation/docs/packages/documentation/routers/nextjs-router.md:709]:
```suggestion
And on the server-side you can use your `accessControlProvider`'s `can` function inside server side functions (`getServerSideProps`) to redirect unauthorized users to other pages..
```

File: documentation/docs/packages/documentation/routers/nextjs-router.md [hunk 8]
```diff
@@ -793,6 +700,175 @@ export const getServerSideProps = async (context) => {
...
+
+:::info
+
+Using the server-side approach is recommended.
```
Comment (aliemir) [inline documentation/docs/packages/documentation/routers/nextjs-router.md:713]:
```suggestion
For page level access control, server-side approach is recommended.
```

Comment (aliemir) [reply documentation/docs/packages/documentation/routers/nextjs-router.md:713]: Also we should have this on Remix as well 😅

File: documentation/docs/migration-guide/router-provider.md [hunk 9]
```diff
@@ -64,9 +64,21 @@ If you have customized the use of `useMenu` hook, you might need to check the us
...
+
+#### Authentication and Access Control
+
+This also means that the authentication and access control redirections are handled according to your framework.
+
+However **refine** still provides some helpers for user's convenience.
+
+For **authentication**, you can use [`Authenticated`](/docs/api-reference/core/components/auth/authenticated) component as wrapper or use `useIsAuthenticated` hook inside your components.
+
+For **authorization**, you can use [`CanAccess`](/docs/api-reference/core/components/accessControl/canAccess) components as wrapper or use `useCan` hook in your components.
```
Comment (aliemir) [inline documentation/docs/migration-guide/router-provider.md]: is this the one link causing error on build?

File: documentation/docs/migration-guide/router-provider.md [hunk 10]
```diff
@@ -64,9 +64,21 @@ If you have customized the use of `useMenu` hook, you might need to check the us
...
+Since **refine** doesn't create routes internally anymore, you are free to create your routes according to your framework without any limitation.
 
-You can find more information and examples in the documentation for each router.
+This means, tasks like `authentication` and `authorization` are also de-coupled from **refine**.
+
+However, for convenience we still provide a set of helpers for you to handle these tasks easily.
+
+#### Authentication and Access Control
+
+This also means that the authentication and access control redirections are handled according to your framework.
```
Comment (aliemir) [inline documentation/docs/migration-guide/router-provider.md:75]:
```suggestion
This also means that the authentication and access control flows should be handled according to your framework.
```

File: documentation/docs/api-reference/core/providers/accessControl-provider.md [hunk 11]
```diff
@@ -62,12 +62,12 @@ const App: React.FC = () => {
...
                 // }
 
-                return Promise.resolve({ can: true });
+                return { can: true };
             },
         }}
     >
         {/* your app */}
-    </Refine>
+    </Refine>;
```
Comment (aliemir) [inline documentation/docs/api-reference/core/providers/accessControl-provider.md:70]: Not a mistake made in this PR but this code block has no return statement in the `App` component. can we wrap the `<Refine>` to parenthesis and a `return` 😅

File: documentation/docs/api-reference/core/components/accessControl/canAccess.md [hunk 12]
```diff
@@ -18,21 +18,46 @@ Passes the given properties to the `can` method from your [access control provid
 
 ## Basic Usage
 
+By default, `CanAccess` component will infer current resource and action based on your route automatically.
```
Comment (aliemir) [inline documentation/docs/api-reference/core/components/accessControl/canAccess.md:21]:
```suggestion
By default, `CanAccess` component will infer the current `resource` and the `action` based on your route automatically. `id` will also be inferred if the current route includes one.
```

File: documentation/docs/advanced-tutorials/access-control.md [hunk 13]
```diff
@@ -249,7 +249,11 @@ const App: React.FC = () => {
...
                         const enforcer = await newEnforcer(model, adapter);
                         // highlight-next-line
-                        const can = await enforcer.enforce(role, resource, action);
+                        const can = await enforcer.enforce(
+                            role,
+                            resource,
+                            action,
+                        );
 
                         return Promise.resolve({
```
Comment (aliemir) [inline documentation/docs/advanced-tutorials/access-control.md:258]: Can we also remove the `Promise.resolve`?

**Pull request merged**